File: rhythm_engine/app/schemas.py
from pydantic import BaseModel, Field
from typing import List, Literal, Union

# Model for standard HTTP errors, used in `responses` documentation
class HTTPError(BaseModel):
    detail: str = Field(..., example="Resource not found")

# --- Reusable Components for Analysis Result ---

class MelodyNote(BaseModel):
    pitch: str = Field(..., description="The MIDI note name (e.g., C4, F#5).", example="A4")
    start_time: float = Field(..., description="Start time of the note in seconds.", example=1.23)
    duration: float = Field(..., description="Duration of the note in seconds.", example=0.45)

class AnalysisInfo(BaseModel):
    total_beats: int = Field(..., description="Total number of beats detected.", example=150)
    total_subdivisions: int = Field(..., description="Total number of subdivisions analyzed.", example=300)
    consolidated_notes: int = Field(..., description="Number of notes after consolidating consecutive identical notes.", example=100)
    filtered_notes: int = Field(..., description="Number of notes after applying the minimum duration filter.", example=80)
    min_note_duration: float = Field(..., description="The minimum note duration filter used.", example=0.05)
    subdivision_factor: int = Field(..., description="The number of subdivisions per beat.", example=2)

class AnalysisResult(BaseModel):
    bpm: float = Field(..., description="Estimated tempo in beats per minute.", example=120.0)
    beat_timestamps: List[float] = Field(..., description="List of timestamps for each detected beat.", example=[0.0, 0.5, 1.0, 1.5])
    melody_map: List[MelodyNote] = Field(..., description="The extracted melody as a list of notes.")
    analysis_info: AnalysisInfo = Field(..., description="Detailed information about the analysis process.")

# --- API Endpoint Specific Models ---

# POST /analyze
class AnalyzeResponse(BaseModel):
    task_id: str = Field(..., description="A unique ID for the analysis task.", example="a1b2c3d4-e5f6-7890-1234-56789abcdef0")
    backend: Literal["celery", "in-memory"] = Field(..., description="The backend used for processing.", example="in-memory")

# GET /results/{task_id} and SSE events
class SuccessResponse(BaseModel):
    state: Literal["SUCCESS"] = Field("SUCCESS", description="The task completed successfully.")
    result: AnalysisResult

class StatusResponse(BaseModel):
    state: Union[Literal["PENDING"], Literal["PROCESSING"]] = Field(..., description="The task is still being processed.", example="PROCESSING")
    status: str = Field(..., description="A message describing the current status.", example="Processing audio file...")

class FailureResponse(BaseModel):
    state: Literal["FAILURE"] = Field("FAILURE", description="The task failed during execution.")
    status: str = Field(..., description="A message describing the error.", example="Audio analysis failed due to an invalid file format.")

# Union model for the /results endpoint
TaskResultResponse = Union[SuccessResponse, StatusResponse, FailureResponse]

# The following models are for documenting the SSE stream, which has additional states not returned by the /results endpoint.
class NotFoundResponse(BaseModel):
    state: Literal["NOT_FOUND"] = Field("NOT_FOUND", description="The requested task ID was not found.")
    status: str = Field("Task not found", description="A message indicating the task ID is not valid.")

class StreamErrorResponse(BaseModel):
    state: Literal["ERROR"] = Field("ERROR", description="An error occurred with the SSE stream itself.")
    status: str = Field(..., example="Stream error: connection closed unexpectedly.") 

File: api/types/schemas.ts
import { z } from "../deps.ts";

// Core audio analysis schemas
export const MelodyNoteSchema = z.object({
  pitch: z.string().describe("MIDI note name (e.g., A4, C#5)"),
  start_time: z.number().describe("Start time in seconds"),
  duration: z.number().describe("Duration in seconds"),
});

export const AnalysisInfoSchema = z.object({
  total_beats: z.number().describe("Total number of beats detected"),
  total_subdivisions: z.number().describe("Total number of subdivisions analyzed"),
  consolidated_notes: z.number().describe("Number of notes after consolidating consecutive identical notes"),
  filtered_notes: z.number().describe("Number of notes after applying the minimum duration filter"),
  min_note_duration: z.number().describe("The minimum note duration filter used"),
  subdivision_factor: z.number().describe("The number of subdivisions per beat"),
});

export const AnalysisResultSchema = z.object({
  bpm: z.number().describe("Estimated tempo in beats per minute"),
  beat_timestamps: z.array(z.number()).describe("List of timestamps for each detected beat"),
  melody_map: z.array(MelodyNoteSchema).describe("The extracted melody as a list of notes"),
  analysis_info: AnalysisInfoSchema.describe("Detailed information about the analysis process"),
});

// API response schemas
export const AnalyzeResponseSchema = z.object({
  task_id: z.string().describe("A unique ID for the analysis task"),
  backend: z.enum(["redis", "in-memory"]).describe("The backend used for processing"),
});

export const SuccessResponseSchema = z.object({
  state: z.literal("SUCCESS"),
  result: AnalysisResultSchema,
});

export const StatusResponseSchema = z.object({
  state: z.enum(["PENDING", "PROCESSING"]),
  status: z.string().describe("A message describing the current status"),
});

export const FailureResponseSchema = z.object({
  state: z.literal("FAILURE"),
  status: z.string().describe("A message describing the error"),
});

export const NotFoundResponseSchema = z.object({
  state: z.literal("NOT_FOUND"),
  status: z.string().default("Task not found"),
});

export const ErrorResponseSchema = z.object({
  state: z.literal("ERROR"),
  status: z.string().describe("Error description"),
});

export const TaskResultResponseSchema = z.union([
  SuccessResponseSchema,
  StatusResponseSchema,
  FailureResponseSchema,
  NotFoundResponseSchema,
  ErrorResponseSchema,
]);

export const HTTPErrorSchema = z.object({
  detail: z.string().describe("Error description"),
});

// Configuration schema
export const ConfigSchema = z.object({
  port: z.number().default(8000),
  log_level: z.enum(["DEBUG", "INFO", "WARN", "ERROR"]).default("INFO"),
  redis_url: z.string().optional(),
  min_note_duration: z.number().default(0.05),
  python_executable: z.string().default("python3"),
  rhythm_engine_path: z.string().default("../rhythm_engine/run.py"),
  upload_dir: z.string().default("./uploads"),
  max_file_size: z.number().default(50 * 1024 * 1024), // 50MB
});

// TypeScript types derived from schemas
export type MelodyNote = z.infer<typeof MelodyNoteSchema>;
export type AnalysisInfo = z.infer<typeof AnalysisInfoSchema>;
export type AnalysisResult = z.infer<typeof AnalysisResultSchema>;
export type AnalyzeResponse = z.infer<typeof AnalyzeResponseSchema>;
export type SuccessResponse = z.infer<typeof SuccessResponseSchema>;
export type StatusResponse = z.infer<typeof StatusResponseSchema>;
export type FailureResponse = z.infer<typeof FailureResponseSchema>;
export type NotFoundResponse = z.infer<typeof NotFoundResponseSchema>;
export type ErrorResponse = z.infer<typeof ErrorResponseSchema>;
export type TaskResultResponse = z.infer<typeof TaskResultResponseSchema>;
export type HTTPError = z.infer<typeof HTTPErrorSchema>;
export type Config = z.infer<typeof ConfigSchema>;

// Supported audio MIME types
export const SUPPORTED_AUDIO_TYPES = [
  "audio/mpeg",
  "audio/wav", 
  "audio/x-wav",
  "audio/mp3",
] as const;

export type SupportedAudioType = typeof SUPPORTED_AUDIO_TYPES[number];